File: app/src/features/notes/services/note.service.ts
/**
 * Note Service
 * 
 * Business logic layer for note operations
 */

import { NoteRepository } from '@/core/database/repositories'
import {
    Note,
    CreateNoteDto,
    UpdateNoteDto,
    NoteStats,
    CreateNoteDtoSchema,
    UpdateNoteDtoSchema,
    builtInNoteTemplates,
    NoteTemplate
} from '../types/note.types'

const noteRepository = new NoteRepository()

export class NoteService {
    /**
     * Create a new note
     */
    static async create(data: CreateNoteDto): Promise<Note> {
        // Validate input
        const validated = CreateNoteDtoSchema.parse(data)
        return await noteRepository.create(validated)
    }

    /**
     * Create a note from template
     */
    static async createFromTemplate(
        projectId: string,
        templateId: string,
        title: string
    ): Promise<Note> {
        const template = builtInNoteTemplates.find(t => t.id === templateId)
        if (!template) {
            throw new Error(`Template '${templateId}' not found`)
        }

        return await this.create({
            projectId,
            title,
            content: template.content,
            tags: template.tags,
            isPinned: false,
        })
    }

    /**
     * Get all notes for a project
     */
    static async getAll(projectId: string): Promise<Note[]> {
        return await noteRepository.findByProjectId(projectId)
    }

    /**
     * Get pinned notes
     */
    static async getPinned(projectId: string): Promise<Note[]> {
        return await noteRepository.findPinned(projectId)
    }

    /**
     * Get recent notes
     */
    static async getRecent(projectId: string, limit: number = 10): Promise<Note[]> {
        return await noteRepository.findRecent(projectId, limit)
    }

    /**
     * Get note by ID
     */
    static async getById(id: string): Promise<Note | null> {
        return await noteRepository.findById(id)
    }

    /**
     * Update note
     */
    static async update(id: string, data: UpdateNoteDto): Promise<Note> {
        // Validate input
        const validated = UpdateNoteDtoSchema.parse(data)
        return await noteRepository.update(id, validated)
    }

    /**
     * Delete note
     */
    static async delete(id: string): Promise<void> {
        return await noteRepository.delete(id)
    }

    /**
     * Toggle pin status
     */
    static async togglePin(id: string): Promise<Note> {
        return await noteRepository.togglePin(id)
    }

    /**
     * Duplicate note
     */
    static async duplicate(id: string, newTitle?: string): Promise<Note> {
        return await noteRepository.duplicate(id, newTitle)
    }

    /**
     * Search notes
     */
    static async search(projectId: string, query: string): Promise<Note[]> {
        return await noteRepository.search(projectId, query)
    }

    /**
     * Get all tags for a project
     */
    static async getAllTags(projectId: string): Promise<string[]> {
        return await noteRepository.getAllTags(projectId)
    }

    /**
     * Get notes by tags
     */
    static async getByTags(projectId: string, tags: string[]): Promise<Note[]> {
        return await noteRepository.findByTags(projectId, tags)
    }

    /**
     * Get note statistics for a project
     */
    static async getStats(projectId: string): Promise<NoteStats> {
        const allNotes = await noteRepository.findByProjectId(projectId)
        const oneWeekAgo = new Date()
        oneWeekAgo.setDate(oneWeekAgo.getDate() - 7)

        return {
            total: allNotes.length,
            pinned: allNotes.filter(n => n.isPinned).length,
            recentlyModified: allNotes.filter(n =>
                new Date(n.updatedAt) > oneWeekAgo
            ).length,
        }
    }

    /**
     * Get available templates
     */
    static getTemplates(): NoteTemplate[] {
        return builtInNoteTemplates
    }

    /**
     * Export note as markdown
     */
    static async exportAsMarkdown(id: string): Promise<string> {
        const note = await noteRepository.findById(id)
        if (!note) {
            throw new Error('Note not found')
        }

        // TODO: Convert Tiptap JSON to Markdown
        // For now, return raw content
        return `# ${note.title}\n\n${note.content}`
    }

    /**
     * Update note content (auto-save friendly)
     */
    static async updateContent(id: string, content: string): Promise<Note> {
        return await noteRepository.update(id, { content })
    }

    /**
     * Add tag to note
     */
    static async addTag(id: string, tag: string): Promise<Note> {
        const note = await noteRepository.findById(id)
        if (!note) {
            throw new Error('Note not found')
        }

        const currentTags = note.tags || []
        if (currentTags.includes(tag)) {
            return note // Tag already exists
        }

        return await noteRepository.update(id, {
            tags: [...currentTags, tag]
        })
    }

    /**
     * Remove tag from note
     */
    static async removeTag(id: string, tag: string): Promise<Note> {
        const note = await noteRepository.findById(id)
        if (!note) {
            throw new Error('Note not found')
        }

        const currentTags = note.tags || []
        return await noteRepository.update(id, {
            tags: currentTags.filter(t => t !== tag)
        })
    }
}

File: app/src-tauri/src/services/note_service.rs
use crate::error::{AppError, AppResult};
use crate::models::{CreateNoteDto, Note, UpdateNoteDto};
use uuid::Uuid;

/// Note service for business logic
pub struct NoteService;

impl NoteService {
    /// Create a new note
    pub async fn create_note(data: CreateNoteDto) -> AppResult<Note> {
        // Validate input
        if data.title.is_empty() {
            return Err(AppError::InvalidInput("Note title cannot be empty".into()));
        }

        if data.project_id.is_empty() {
            return Err(AppError::InvalidInput("Project ID cannot be empty".into()));
        }

        let now = chrono::Utc::now().timestamp();
        let note = Note {
            id: Uuid::new_v4().to_string(),
            project_id: data.project_id,
            title: data.title,
            content: data.content,
            created_at: now,
            updated_at: now,
            tags: data.tags,
            is_pinned: data.is_pinned.unwrap_or(false),
        };

        // TODO: Save to database via IPC to frontend repository

        Ok(note)
    }

    /// Get all notes for a project
    pub async fn list_notes(project_id: String) -> AppResult<Vec<Note>> {
        if project_id.is_empty() {
            return Err(AppError::InvalidInput("Project ID cannot be empty".into()));
        }

        // TODO: Fetch from database
        Ok(vec![])
    }

    /// Get pinned notes for a project
    pub async fn list_pinned_notes(project_id: String) -> AppResult<Vec<Note>> {
        if project_id.is_empty() {
            return Err(AppError::InvalidInput("Project ID cannot be empty".into()));
        }

        // TODO: Fetch from database
        Ok(vec![])
    }

    /// Get recent notes
    pub async fn list_recent_notes(project_id: String, limit: i32) -> AppResult<Vec<Note>> {
        if project_id.is_empty() {
            return Err(AppError::InvalidInput("Project ID cannot be empty".into()));
        }

        if limit <= 0 {
            return Err(AppError::InvalidInput("Limit must be greater than 0".into()));
        }

        // TODO: Fetch from database with limit
        Ok(vec![])
    }

    /// Get note by ID
    pub async fn get_note(id: String) -> AppResult<Note> {
        if id.is_empty() {
            return Err(AppError::InvalidInput("Note ID cannot be empty".into()));
        }

        // TODO: Fetch from database
        Err(AppError::NotFound("Note", id))
    }

    /// Update note
    pub async fn update_note(id: String, _data: UpdateNoteDto) -> AppResult<Note> {
        if id.is_empty() {
            return Err(AppError::InvalidInput("Note ID cannot be empty".into()));
        }

        // TODO: Update in database
        Err(AppError::NotFound("Note", id))
    }

    /// Delete note
    pub async fn delete_note(id: String) -> AppResult<()> {
        if id.is_empty() {
            return Err(AppError::InvalidInput("Note ID cannot be empty".into()));
        }

        // TODO: Delete from database
        Err(AppError::NotFound("Note", id))
    }

    /// Toggle pin status
    pub async fn toggle_pin(id: String) -> AppResult<Note> {
        if id.is_empty() {
            return Err(AppError::InvalidInput("Note ID cannot be empty".into()));
        }

        // TODO: Toggle pin in database
        Err(AppError::NotFound("Note", id))
    }

    /// Duplicate note
    pub async fn duplicate_note(id: String, _new_title: Option<String>) -> AppResult<Note> {
        if id.is_empty() {
            return Err(AppError::InvalidInput("Note ID cannot be empty".into()));
        }

        // TODO: Duplicate in database
        Err(AppError::NotFound("Note", id))
    }

    /// Search notes
    pub async fn search_notes(project_id: String, _query: String) -> AppResult<Vec<Note>> {
        if project_id.is_empty() {
            return Err(AppError::InvalidInput("Project ID cannot be empty".into()));
        }

        // TODO: Search in database
        Ok(vec![])
    }

    /// Get all tags for a project
    pub async fn get_all_tags(project_id: String) -> AppResult<Vec<String>> {
        if project_id.is_empty() {
            return Err(AppError::InvalidInput("Project ID cannot be empty".into()));
        }

        // TODO: Fetch unique tags from database
        Ok(vec![])
    }

    /// Get notes by tags
    pub async fn list_notes_by_tags(project_id: String, tags: Vec<String>) -> AppResult<Vec<Note>> {
        if project_id.is_empty() {
            return Err(AppError::InvalidInput("Project ID cannot be empty".into()));
        }

        if tags.is_empty() {
            return Err(AppError::InvalidInput("Tags cannot be empty".into()));
        }

        // TODO: Fetch from database filtered by tags
        Ok(vec![])
    }
}
